.github/workflows/behave.yml

```diff
@@ -17,63 +17,71 @@ jobs:
   behave:
     name: Run Behave
     runs-on: ubuntu-latest
-
-    services:
-      postgres:
-        image: postgres:latest
-        env:
-          POSTGRES_USER: scram
-          POSTGRES_PASSWORD: ''
-          POSTGRES_DB: test_scram
-          POSTGRES_HOST_AUTH_METHOD: trust
-        ports:
-          - 5432:5432
-        options: >-
-          --health-cmd "pg_isready -U scram"
-          --health-interval 10s
-          --health-timeout 5s
-          --health-retries 5
+    strategy:
+      max-parallel: 4
+      matrix:
+        python-version: ['3.11', '3.12']
 
     steps:
       - name: Check out the code
         uses: actions/checkout@v4
 
       - name: Set up Docker
         uses: docker/setup-buildx-action@v3
+
       - name: Install Docker Compose
         run: |
           sudo apt-get update
           sudo apt-get install -y docker-compose make
 
+      - name: Check Docker state (pre-build)
+        run: docker ps
+
       - name: Build Docker images
         run: make build
+        env:
+          PYTHON_IMAGE_VER: "${{ matrix.python-version }}"
 
       - name: Migrate Database
         run: make migrate
 
       - name: Run Application
         run: make run
 
+      - name: Check Docker state (pre-test)
+        run: docker ps
+
       - name: Run pytest + behave with Coverage
         env:
           POSTGRES_USER: scram
           POSTGRES_DB: test_scram
         run: make coverage.xml
 
+      - name: Dump docker logs on failure
+        if: failure()
+        uses: jwalton/gh-docker-logs@v2
+
       - name: Upload Coverage to Coveralls
+        if: matrix.python-version == '3.12'
         uses: coverallsapp/github-action@v2
 
       - name: Upload Coverage to GitHub
+        if: matrix.python-version == '3.12'
         uses: actions/upload-artifact@v4
         with:
           name: coverage-report
           path: coverage.xml
 
       - name: Display Coverage Metrics
+        if: matrix.python-version == '3.12'
```

Comment:
should/can we variablize this?

Reply:
I don't think there's a way to define the "last" element of the matrix. But it was failing the CI run if we tried to upload more than once.

Reply:
Ah ok yeah I just looked and all of the solutions look even uglier than this and still require repeating yourself a lot. this works